File: src/src/Systems/PopulationSystem.cs
using Godot;
using System;
using System.Collections.Generic;

public class PopulationSystem : Node, IUpdatable
{
    private GameManager gameManager;
    private Random rnd = new Random();

    // Prefabs (PackedScene instances)
    private PackedScene treePrefab;
    private PackedScene buildingPrefab;

    public PopulationSystem(GameManager manager)
    {
        gameManager = manager;
    }

    public override void _Ready()
    {
        treePrefab = GD.Load<PackedScene>("res://src/Graphics/Models/Trees/Tree.tscn");
        buildingPrefab = GD.Load<PackedScene>("res://src/Graphics/Models/Buildings/Building.tscn");
    }

    public void Update(float deltaTime)
    {
        // Example: Populate areas periodically
        // You can tie this to game time or specific events

        // For demonstration, let's populate every 60 seconds
        // Implement a timer or counter as needed
    }

    public void PopulateArea(Vector3 position, string type)
    {
        if (type == "tree")
        {
            var tree = (MeshInstance)treePrefab.Instance();
            tree.Transform = new Transform(Basis.Identity, position + new Vector3(rnd.Next(-5, 5), 0, rnd.Next(-5, 5)));
            AddChild(tree);
        }
        else if (type == "building")
        {
            var building = (MeshInstance)buildingPrefab.Instance();
            building.Transform = new Transform(Basis.Identity, position + new Vector3(rnd.Next(-10, 10), 0, rnd.Next(-10, 10)));
            AddChild(building);
        }
    }

    // Example method to populate multiple areas
    public void PopulateMultipleAreas()
    {
        foreach (var node in gameManager.Nodes)
        {
            // Populate trees around each node
            for (int i = 0; i < 10; i++)
            {
                PopulateArea(node.Position, "tree");
            }

            // Populate buildings around each node
            for (int i = 0; i < 5; i++)
            {
                PopulateArea(node.Position, "building");
            }
        }
    }
}

File: src/Systems/PopulationSystem.cpp
#include "PopulationSystem.hpp"
#include <ResourceLoader.hpp>
#include <MeshInstance.hpp>
#include <Transform.hpp>
#include <Basis.hpp>

namespace Systems {

using namespace godot;

void PopulationSystem::_register_methods() {
    register_method("_ready", &PopulationSystem::_ready);
    register_method("update", &PopulationSystem::update);
    register_method("populate_area", &PopulationSystem::populate_area);
    register_method("populate_multiple_areas", &PopulationSystem::populate_multiple_areas);
}

PopulationSystem::PopulationSystem() {}
PopulationSystem::~PopulationSystem() {}

void PopulationSystem::_init() {
    rng.instance();
    rng->randomize();
}

void PopulationSystem::_ready() {
    tree_prefab = ResourceLoader::get_singleton()->load("res://src/Graphics/Models/Trees/Tree.tscn");
    building_prefab = ResourceLoader::get_singleton()->load("res://src/Graphics/Models/Buildings/Building.tscn");
}

void PopulationSystem::update(float delta_time) {
    // Implement periodic population updates
}

void PopulationSystem::populate_area(const Vector3& position, const String& type) {
    Ref<PackedScene> prefab;
    Vector3 offset_range;
    
    if (type == "tree") {
        prefab = tree_prefab;
        offset_range = Vector3(5, 0, 5);
    } else if (type == "building") {
        prefab = building_prefab;
        offset_range = Vector3(10, 0, 10);
    } else {
        return;
    }

    if (prefab.is_valid()) {
        Node* instance = prefab->instance();
        MeshInstance* mesh_instance = Object::cast_to<MeshInstance>(instance);
        
        if (mesh_instance) {
            Vector3 random_offset(
                rng->randf_range(-offset_range.x, offset_range.x),
                0,
                rng->randf_range(-offset_range.z, offset_range.z)
            );
            
            Transform transform;
            transform.set_basis(Basis());
            transform.set_origin(position + random_offset);
            
            mesh_instance->set_transform(transform);
            add_child(mesh_instance);
        }
    }
}

void PopulationSystem::populate_multiple_areas() {
    // Implementation depends on your GameManager structure
    // Example:
    /*
    for (const auto& node : game_manager->get_nodes()) {
        // Populate trees
        for (int i = 0; i < 10; i++) {
            populate_area(node->get_position(), "tree");
        }
        
        // Populate buildings
        for (int i = 0; i < 5; i++) {
            populate_area(node->get_position(), "building");
        }
    }
    */
}

} // namespace Systems 